storage/internal/benchmarks/main.go

```diff
@@ -0,0 +1,367 @@
+	objectSize    int64
+	appBufferSize int
+	chunkSize     int
+	crc32Enabled  bool
+	md5Enabled    bool
+	API           benchmarkAPI
+	elapsedTime   time.Duration
+	completed     bool
+	isRead        bool
+	readIteration int
+	heapSys       uint64
+	heapAlloc     uint64
+	stackInUse    uint64
+	heapAllocDiff uint64
+	mallocsDiff   uint64
+	start         time.Time
+}
+
+func benchmarkRun(ctx context.Context, opts *benchmarkOptions, bucketName string) error {
+	var memStats *runtime.MemStats = &runtime.MemStats{}
+
+	_, doMD5, doCRC32C := randomOf3()
+
+	objectName := randomName("obj")
+	objectSize := randomInt64(opts.minObjectSize, opts.maxObjectSize)
+
+	appWriteBufferSize := opts.writeQuantum * randomInt(opts.minWriteSize/opts.writeQuantum, opts.maxWriteSize/opts.writeQuantum)
+	appReadBufferSize := opts.readQuantum * randomInt(opts.minReadSize/opts.readQuantum, opts.maxReadSize/opts.readQuantum)
+
+	writeChunkSize := randomInt64(opts.minChunkSize, opts.maxChunkSize)
+
+	// Select client
+	client, readAPI, writeAPI, err := initializeClient(ctx, opts.api, appWriteBufferSize, appReadBufferSize, opts.connPoolSize)
+	if err != nil {
+		return fmt.Errorf("NewClient: %v", err)
+	}
+
+	// Create contents to write
+	err = generateRandomFile(objectName, objectSize)
+	if err != nil {
+		return fmt.Errorf("generateRandomFile: %v", err)
+	}
+
+	o := client.Bucket(bucketName).Object(objectName)
+	defer o.Delete(ctx)
+
+	// Upload
+
+	// If the option is specified, run a garbage collector before collecting
+	// memory statistics and starting the timer on the benchmark. This can be
+	// used to compare between running each benchmark "on a blank slate" vs organically.
+	forceGarbageCollection(opts.forceGC)
+
+	runtime.ReadMemStats(memStats)
+	prevHeapAlloc := memStats.HeapAlloc
+	prevMallocs := memStats.Mallocs
+
+	start := time.Now()
+	timeTaken, err := uploadBenchmark(ctx, uploadOpts{
+		o:         o,
+		fileName:  objectName,
+		chunkSize: int(writeChunkSize),
+		md5:       doMD5,
+		crc32c:    doCRC32C,
+	})
+	runtime.ReadMemStats(memStats)
+	results <- benchmarkResult{
+		objectSize:    objectSize,
+		appBufferSize: appWriteBufferSize,
+		chunkSize:     int(writeChunkSize),
+		crc32Enabled:  doCRC32C,
+		md5Enabled:    doMD5,
+		API:           readAPI,
+		elapsedTime:   timeTaken,
+		completed:     err == nil,
+		isRead:        false,
+		heapSys:       memStats.HeapSys,
+		heapAlloc:     memStats.HeapAlloc,
+		stackInUse:    memStats.StackInuse,
+		heapAllocDiff: memStats.HeapAlloc - prevHeapAlloc,
+		mallocsDiff:   memStats.Mallocs - prevMallocs,
+		start:         start,
+	}
+	os.Remove(objectName)
+	// Do not attempt to read from a failed upload
+	if err != nil {
+		return fmt.Errorf("failed upload: %v", err)
+	}
+
+	// Wait for the object to be available.
+	timedCtx, cancelTimedCtx := context.WithTimeout(ctx, time.Second*3)
+	defer cancelTimedCtx()
+	if _, err := o.Retryer(storage.WithPolicy(storage.RetryAlways)).Attrs(timedCtx); err != nil {
```

Comment:
Not sure I understand this line; but does it always retry for any failures on subsequent storage requests? (which in this case would be reads which are always retriable)